File: assets/manual-login.mjs
#!/usr/bin/env node
/**
 * 手动登录并导出登录态
 *
 * 站点在风控触发后会要求滑块验证码，自动登录就走不通了。验证码本来就是用来
 * 拦自动化的，不该去绕；正确做法是让人过一次，然后复用会话 —— 登录态能用一段时间，
 * 期间所有云端操作都不必再登录。
 *
 * 用法：node manual-login.mjs
 *   浏览器窗口打开 → 你手动登录（含验证码）→ 检测到登录成功后自动导出并关闭
 */

import { chromium } from '@playwright/test';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { resolveChrome } from '../scripts/chrome-path.mjs';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const BASE = process.env.REC_BASE_URL;
if (!BASE) { console.error('请设置 REC_BASE_URL，或改用 config.json'); process.exit(1); }
const ENTRY = BASE + (process.env.REC_ENTRY_PATH ?? '/');
const AUTH = path.join(__dirname, '.auth');
const WAIT_MINUTES = 15;

const browser = await chromium.launch({
  headless: false,
  executablePath: resolveChrome(),
  args: ['--ignore-certificate-errors', '--start-maximized'],
});
const ctx = await browser.newContext({ ignoreHTTPSErrors: true, viewport: null });
const page = await ctx.newPage();
page.setDefaultNavigationTimeout(90_000);

console.log(`\n浏览器已打开：${ENTRY}`);
console.log('请在窗口里登录（含滑块验证码）。检测到登录成功后会自动导出并关闭窗口。');
console.log(`最多等待 ${WAIT_MINUTES} 分钟。\n`);

await page.goto(ENTRY, { waitUntil: 'domcontentloaded' }).catch(() => {});

// 判据用应用侧状态而不是 URL —— SSO 回调会多跳几次，参数还不固定。
// .catch 是必须的：轮询期间撞上导航，evaluate 会抛 Execution context was destroyed。
const deadline = Date.now() + WAIT_MINUTES * 60_000;
let ok = false;
while (Date.now() < deadline) {
  ok = await page
    .evaluate(() => sessionStorage.getItem('userInfo') !== null)
    .catch(() => false);
  if (ok) break;
  await page.waitForTimeout(1500);
}

if (!ok) {
  console.error(`\n❌ ${WAIT_MINUTES} 分钟内未检测到登录成功，未导出任何内容。`);
  await browser.close().catch(() => {});
  process.exit(1);
}

// 关掉登录后的一次性弹窗，避免它出现在后续用例里
const dlg = page.getByRole('dialog');
if (await dlg.isVisible().catch(() => false)) {
  await dlg.getByRole('button', { name: /关闭|×|我知道了/ }).first().click().catch(() => {});
}
await page.waitForTimeout(1500);

fs.mkdirSync(AUTH, { recursive: true });
await ctx.storageState({ path: path.join(AUTH, 'state.json') });
fs.writeFileSync(path.join(AUTH, 'session-storage.json'),
  await page.evaluate(() => JSON.stringify(sessionStorage)), 'utf-8');
fs.writeFileSync(path.join(AUTH, 'cookies.json'),
  JSON.stringify(await ctx.cookies(), null, 2), 'utf-8');

// 立刻验证导出的凭据真能调通接口。只看到"登录成功"是不够的 ——
// 否则后面每个用例都会因为同一个原因失败，而失败点散落各处。
// 自检用的只读接口，按站点改；留空则只探首页
const PROBE = process.env.REC_PROBE_PATH ?? '/';
const probe = await page.evaluate(async (u) => {
  const r = await fetch(u);
  return { status: r.status, json: (r.headers.get('content-type') ?? '').includes('json') };
}, PROBE);

console.log(`\n✅ 登录态已导出到 ${AUTH}/`);
console.log(`   连通性自检：HTTP ${probe.status}，返回 JSON = ${probe.json}`);
if (!probe.json) console.log('   ⚠ 返回的不是 JSON，会话可能仍有问题');

await browser.close().catch(() => {});

File: assets/manual_login.py
#!/usr/bin/env python3
"""手动登录并导出登录态（manual-login.mjs 的 Python 版）。

站点在风控触发后会要求滑块验证码，自动登录就走不通了。验证码本来就是用来
拦自动化的，不该去绕；正确做法是让人过一次，然后复用会话 —— 登录态能用一段时间，
期间所有云端操作都不必再登录。

用法：python manual_login.py
  浏览器窗口打开 → 你手动登录（含验证码）→ 检测到登录成功后自动导出并关闭
"""

import os
import sys
import time
from pathlib import Path

from playwright.sync_api import sync_playwright

sys.path.insert(0, str(Path(__file__).resolve().parent))

from auth_setup import dismiss_dialogs, export_state  # noqa: E402
from chrome_path import resolve_chrome                # noqa: E402

HERE = Path(__file__).resolve().parent
AUTH = HERE / ".auth"
WAIT_MINUTES = 15

BASE = os.environ.get("REC_BASE_URL")
if not BASE:
    sys.exit("请设置 REC_BASE_URL")
ENTRY = BASE.rstrip("/") + os.environ.get("REC_ENTRY_PATH", "/")

# 自检用的只读接口，按站点改；留空则只探首页
PROBE = os.environ.get("REC_PROBE_PATH", "/")


def main() -> int:
    with sync_playwright() as p:
        browser = p.chromium.launch(
            headless=False,
            executable_path=resolve_chrome(),
            args=["--ignore-certificate-errors", "--start-maximized"],
        )
        # viewport=None 让页面跟着真实窗口尺寸走 —— 手动操作时锁死 viewport
        # 会把底部的登录按钮挤到可视区外
        ctx = browser.new_context(ignore_https_errors=True, no_viewport=True)
        page = ctx.new_page()
        page.set_default_navigation_timeout(90_000)

        print(f"\n浏览器已打开：{ENTRY}")
        print("请在窗口里登录（含滑块验证码）。检测到登录成功后会自动导出并关闭窗口。")
        print(f"最多等待 {WAIT_MINUTES} 分钟。\n")

        try:
            page.goto(ENTRY, wait_until="domcontentloaded")
        except Exception:
            pass

        # 判据用应用侧状态而不是 URL —— SSO 回调会多跳几次，参数还不固定。
        # try/except 是必须的：轮询期间撞上导航，evaluate 会抛
        # Execution context was destroyed。
        deadline = time.monotonic() + WAIT_MINUTES * 60
        ok = False
        while time.monotonic() < deadline:
            try:
                ok = bool(page.evaluate(
                    "() => sessionStorage.getItem('userInfo') !== null"))
            except Exception:
                ok = False
            if ok:
                break
            page.wait_for_timeout(1500)

        if not ok:
            print(f"\n❌ {WAIT_MINUTES} 分钟内未检测到登录成功，未导出任何内容。")
            browser.close()
            return 1

        dismiss_dialogs(page)
        page.wait_for_timeout(1500)

        export_state(ctx, page, AUTH)

        # 立刻验证导出的凭据真能调通接口。只看到"登录成功"是不够的 ——
        # 否则后面每个用例都会因为同一个原因失败，而失败点散落各处。
        probe = page.evaluate(
            """async (u) => {
                 const r = await fetch(u);
                 return { status: r.status,
                          json: (r.headers.get('content-type') ?? '').includes('json') };
               }""",
            PROBE,
        )
        print(f"\n✅ 登录态已导出到 {AUTH}/")
        print(f"   连通性自检：HTTP {probe['status']}，返回 JSON = {probe['json']}")
        if not probe["json"]:
            print("   ⚠ 返回的不是 JSON，会话可能仍有问题")

        browser.close()
        return 0


if __name__ == "__main__":
    sys.exit(main())
